substratevm/src/com.oracle.svm.test/src/com/oracle/svm/test/jfr/utils/JfrFileParser.java

```diff
@@ -135,49 +131,47 @@ private static long parseConstantPoolHeader(RecordingInput input, long constantP
         // Size of constant pools.
         assertTrue("Constant pool size is invalid!", input.readInt() > 0);
         // Constant pools region ID.
-        assertEquals(JfrReservedEvent.EVENT_CHECKPOINT.getId(), input.readLong());
+        assertEquals(JfrReservedEvent.CHECKPOINT.getId(), input.readLong());
         assertTrue("Constant pool timestamp is invalid!", input.readLong() > 0); // Timestamp.
         input.readLong(); // Duration.
         long deltaNext = input.readLong(); // Offset to a next constant pools region.
         assertTrue(input.readBoolean()); // Flush.
         return deltaNext;
     }
 
-    private static void compareFoundAndExpectedIds() {
-        for (ConstantPoolParser parser : supportedConstantPools.values()) {
-            parser.compareFoundAndExpectedIds();
-        }
-    }
-
-    /**
-     * Must verify constant pools in order that they were written because event streaming can write
-     * pools before the chunk is finished. This means that a given pool may reference constants from
-     * another pool written previously (within the same chunk).
-     */
     private static void verifyConstantPools(RecordingInput input, long constantPoolPosition) throws IOException {
-        List<Long> poolPositions = new ArrayList<>();
         long deltaNext;
         long currentConstantPoolPosition = constantPoolPosition;
         do {
-            poolPositions.add(currentConstantPoolPosition);
             deltaNext = parseConstantPoolHeader(input, currentConstantPoolPosition);
-            currentConstantPoolPosition += deltaNext;
-        } while (deltaNext != 0);
-
-        for (int j = poolPositions.size() - 1; j > 0; j--) {
-            parseConstantPoolHeader(input, poolPositions.get(j));
             long numberOfCPs = input.readInt();
             for (int i = 0; i < numberOfCPs; i++) {
                 ConstantPoolParser constantPoolParser = supportedConstantPools.get(input.readLong());
                 Assert.assertNotNull("Unknown constant pool!", constantPoolParser);
                 constantPoolParser.parse(input);
             }
-            compareFoundAndExpectedIds();
+            currentConstantPoolPosition += deltaNext;
+        } while (deltaNext != 0);
+
+        /* Now that we collected all data, verify and clear it. */
```

Comment:
Would it be better to check with `compareFoundAndExpectedIds()` after each constant pool? This way we can validate each flush individually to make sure there's no unexpected references at any intermediate point?

Reply:
Good point, we can indeed verify the accumulated constant pools after every checkpoint. I changed this so that the code is now similar to what you had in place before.